dev/ProgressRing/ProgressRing.cpp

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "common.h"
+#include "ProgressRing.h"
+#include "RuntimeProfiler.h"
+#include "ResourceAccessor.h"
+#include "math.h"
+
+ProgressRing::ProgressRing()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_ProgressRing);
+
+    SetDefaultStyleKey(this);
+
+    RegisterPropertyChangedCallback(winrt::RangeBase::ValueProperty(), { this, &ProgressRing::OnRangeBasePropertyChanged });
+
+    SizeChanged({ this, &ProgressRing::OnSizeChanged });
+}
+
+void ProgressRing::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected{ *this };
+
+    m_layoutRoot.set(GetTemplateChildT<winrt::Grid>(s_LayoutRootName, controlProtected));
+    m_outlineFigure.set(GetTemplateChildT<winrt::PathFigure>(s_OutlineFigureName, controlProtected));
+    m_outlineArc.set(GetTemplateChildT<winrt::ArcSegment>(s_OutlineArcName, controlProtected));
+    m_barFigure.set(GetTemplateChildT<winrt::PathFigure>(s_BarFigureName, controlProtected));
+    m_barArc.set(GetTemplateChildT<winrt::ArcSegment>(s_BarArcName, controlProtected));
+
+    RenderAll();
+}
+
+void ProgressRing::OnSizeChanged(const winrt::IInspectable&, const winrt::IInspectable&)
+{
+    RenderAll();
+}
+
+void ProgressRing::OnRangeBasePropertyChanged(const winrt::DependencyObject&, const winrt::DependencyProperty&)
+{
+    RenderSegment();
+}
+
+void ProgressRing::OnStrokeThicknessPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    RenderAll();
+}
+
+winrt::Size ProgressRing::ComputeEllipseSize(const double thickness, const double actualWidth, const double actualHeight)
+{
+    const double safeThickness = std::max(thickness, static_cast<double>(0.0));
+    const double width = std::max((actualWidth - safeThickness) / 2.0, 0.0);
+    const double height = std::max((actualHeight - safeThickness) / 2.0, 0.0);
+
+    return {static_cast<float>(width), static_cast<float>(height)};
+}
+
+
+void ProgressRing::RenderSegment()
```

Comment:
>RenderSegment [](start = 19, length = 13)

nit: Render has a different meaning in Xaml, so perhaps name these as Update or Calculate or something along those lines ?